cpp/src/copying/gather_opt.cu

```diff
@@ -0,0 +1,360 @@
+  gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+ 
+  const bool active_threads = destination_row < num_destination_rows;
+  gdf_index_type source_row = active_threads ? gather_map[destination_row] : 0;
+  for(gdf_index_type i = 0; i < destination.num_columns(); i++){
+    if(active_threads && (!check_bounds || source_row < num_source_rows)){
+      // gather the entire row
+      cudf::type_dispatcher(source.get_column(i)->dtype,
+                          copy_element_smem{},
+                          *destination.get_column(i), destination_row,
+                          *source.get_column(i), source_row);
+    }
+    
+    if(do_valid){
+      // Before bit_mask_t is used in device_table we will do the cast here.
+      bit_mask_t* __restrict__ src_valid =
+        reinterpret_cast<bit_mask_t*>(source.get_column(i)->valid);
+      bit_mask_t* __restrict__ dest_valid =
+        reinterpret_cast<bit_mask_t*>(destination.get_column(i)->valid);
+      // Gather the bitmasks and do the null count 
+      gather_bitmask_device<check_bounds>(src_valid, source_row, num_source_rows,
+        dest_valid, destination_row, num_destination_rows, d_count + i);
+    }
+  }
+}
+
+ 
+template<bool check_bounds, bool do_valid, int block_size>
+__launch_bounds__(block_size, 2048/block_size)
+__global__ void gather_kernel_sparse(const device_table source, 
+                                  const gdf_size_type gather_map[], 
+                                  device_table destination,
+                                  gdf_size_type* d_count)
+{
+  const gdf_index_type num_source_rows = source.num_rows();
+  const gdf_index_type num_destination_rows = destination.num_rows();
+  for(gdf_index_type i = 0; i < destination.num_columns(); i++){
+    
+    // data part
+    gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+    while(destination_row < num_destination_rows){
+      
+      gdf_index_type source_row = gather_map[destination_row];
+      
+      cudf::type_dispatcher(source.get_column(i)->dtype,
+                            copy_element_smem{},
+                            *destination.get_column(i), destination_row,
+                            *source.get_column(i), source_row);
+      
+      destination_row += blockDim.x * gridDim.x; 
+    }
+    if(do_valid){ 
+      // bitmask part  
+      gdf_index_type destination_row_base = blockIdx.x * blockDim.x;
+      while(destination_row_base < num_destination_rows){
+        
+        gdf_index_type destination_row = destination_row_base + threadIdx.x;
+        const bool active_threads = destination_row < num_destination_rows;
+        gdf_index_type source_row = active_threads ? gather_map[destination_row] : 0;
+        
+        // Before bit_mask_t is used in device_table we will do the cast here.
+        bit_mask_t* __restrict__ src_valid =
+          reinterpret_cast<bit_mask_t*>(source.get_column(i)->valid);
```

Comment:
Doing `reinterpret_casts` inside kernels like this can mess with the compiler's ability to do optimization because it messes with aliasing.

Perhaps you could experiment with modifying `device_table` to use `bit_mask_t` directly and avoid the cast inside the kernel to see if there is an impact on performance.